File: js/handlers/memoryFilterHandler.js
const { MemoryFilterBuilder } = require('filtex-js');
const { ProjectFilter } = require('../filters/projectFilter');
const { MemoryStorage } = require('../storages/memoryStorage');
const {MemoryExpression} = require("filtex-js/dist/builders/memory/types/memoryExpression");

class MemoryFilterHandler {
    async handleAsync(req, res) {
        try {
            let expression;

            switch (req.query.type) {
                case 'text':
                    if (!req.body.query || req.body.query === '') {
                        break;
                    }

                    ProjectFilter.validateFromText(req.body.query);
                    expression = ProjectFilter.expressionFromText(req.body.query);
                    break;
                case 'json':
                    if (!req.body.query) {
                        break;
                    }

                    ProjectFilter.validateFromJson(JSON.stringify(req.body.query));
                    expression = ProjectFilter.expressionFromJson(JSON.stringify(req.body.query));
                    break;
                default:
                    throw new Error('invalid type');
            }

            let memoryFilter;

            if (expression) {
                memoryFilter = new MemoryFilterBuilder().build(expression);
            } else {
                memoryFilter = new MemoryExpression((_) => true);
            }

            const results = new MemoryStorage().Query(memoryFilter.fn);

            res.status(200).json(results);
        } catch (ex) {
            res.status(500).send(ex.message);
        }
    }

}

module.exports = { MemoryFilterHandler };

File: go/handlers/memory_filter_handler.go
package handlers

import (
	"encoding/json"
	"filtex-go-example/filters"
	"filtex-go-example/storages"
	"github.com/filtex/filtex-go/builders/memory"
	"github.com/filtex/filtex-go/builders/memory/types"
	"github.com/filtex/filtex-go/expressions"
	"net/http"
)

func MemoryFilterHandler(w http.ResponseWriter, r *http.Request) {
	projectFilter, err := filters.ProjectFilter()
	if err != nil {
		http.Error(w, "Error generating Filtex instance", http.StatusInternalServerError)
		return
	}

	var requestBody map[string]interface{}
	decoder := json.NewDecoder(r.Body)
	err = decoder.Decode(&requestBody)
	if err != nil {
		http.Error(w, "Error decoding JSON request", http.StatusBadRequest)
		return
	}

	var expression expressions.Expression

	switch r.URL.Query().Get("type") {
	case "text":
		if requestBody["query"] == nil {
			break
		}

		query, ok := requestBody["query"].(string)
		if !ok {
			http.Error(w, "Invalid request format", http.StatusBadRequest)
			return
		}

		if query == "" {
			break
		}

		err = projectFilter.ValidateFromText(query)
		if err != nil {
			http.Error(w, "invalid query", http.StatusBadRequest)
			return
		}

		expression, err = projectFilter.ExpressionFromText(query)
		if err != nil {
			http.Error(w, "invalid query", http.StatusBadRequest)
			return
		}
	case "json":
		if requestBody["query"] == nil {
			break
		}

		query, err := json.Marshal(requestBody["query"])
		if err != nil {
			http.Error(w, "Invalid request format", http.StatusBadRequest)
			return
		}

		err = projectFilter.ValidateFromJson(string(query))
		if err != nil {
			http.Error(w, "invalid query", http.StatusBadRequest)
			return
		}

		expression, err = projectFilter.ExpressionFromJson(string(query))
		if err != nil {
			http.Error(w, "invalid query", http.StatusBadRequest)
			return
		}
	default:
		http.Error(w, "invalid type", http.StatusBadRequest)
		return
	}

	var memoryFilter *types.MemoryExpression

	if expression != nil {
		memoryFilter, err = memory.NewMemoryFilterBuilder().Build(expression)
		if err != nil {
			http.Error(w, "invalid query", http.StatusBadRequest)
			return
		}
	} else {
		memoryFilter = &types.MemoryExpression{
			Fn: func(i map[string]interface{}) bool {
				return true
			},
		}
	}

	results, err := storages.MemoryStorage{}.Query(memoryFilter.Fn)
	if err != nil {
		http.Error(w, "could not be fetched", http.StatusBadRequest)
		return
	}

	jsonResponse, err := json.Marshal(results)
	if err != nil {
		http.Error(w, "Error encoding JSON response", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")

	w.Write(jsonResponse)
}
